R/calculateDistance.R

```diff
@@ -1,6 +1,315 @@
+#'   all the abundances of features are equal between two samples, and 0 means 
+#'   that samples have completely different relative abundances. 
+#'   
+#'   Mia unifrac internal function utilizes 
+#'   \code{\link[rbiom:unifrac]{rbiom:unifrac()}}.
+#'   
+#' @name getDissimilarity
+#' 
+#' @author
+#' For overlap implementation: 
+#' Leo Lahti and Tuomas Borman. Contact: \url{microbiome.github.io}
+#' 
+#' For JSD implementation:
+#' Susan Holmes \email{susan@@stat.stanford.edu}.
+#' Adapted for phyloseq by Paul J. McMurdie.
+#' Adapted for mia by Felix G.M. Ernst
+#' 
+#' @seealso
+#' \url{http://en.wikipedia.org/wiki/Jensen-Shannon_divergence}
+#' 
+#' @references
+#' For unifrac dissimilarity: \url{http://bmf.colorado.edu/unifrac/}
+#' 
+#' See also additional descriptions of Unifrac in the following articles:
+#'
+#' Lozupone, Hamady and Knight, ``Unifrac - An Online Tool for Comparing
+#' Microbial Community Diversity in a Phylogenetic Context.'', BMC
+#' Bioinformatics 2006, 7:371
+#'
+#' Lozupone, Hamady, Kelley and Knight, ``Quantitative and qualitative (beta)
+#' diversity measures lead to different insights into factors that structure
+#' microbial communities.'' Appl Environ Microbiol. 2007
+#'
+#' Lozupone C, Knight R. ``Unifrac: a new phylogenetic method for comparing
+#' microbial communities.'' Appl Environ Microbiol. 2005 71 (12):8228-35.
+#' 
+#' For JSD dissimilarity: 
+#' Jensen-Shannon Divergence and Hilbert space embedding.
+#' Bent Fuglede and Flemming Topsoe University of Copenhagen,
+#' Department of Mathematics
+#' \url{http://www.math.ku.dk/~topsoe/ISIT2004JSD.pdf}
+#'
+#' @export
+#'
+#' @examples
+#' library(mia)
+#' library(scater)
+#' 
+#' # load dataset
+#' data(GlobalPatterns)
+#' tse <- GlobalPatterns
+#' 
+#' ### Overlap dissimilarity
+#' 
+#' tse <- addDissimilarity(tse, method = "overlap", detection = 0.25)
+#' reducedDim(tse, "overlap")[1:6, 1:6]
+#' 
+#' ### JSD dissimilarity
+#' 
+#' tse <- addDissimilarity(tse, method = "jsd")
+#' reducedDim(tse, "jsd")[1:6, 1:6]
+#' 
+#' # Multi Dimensional Scaling applied to JSD distance matrix
+#' tse <- runMDS(tse, FUN = getDissimilarity, method = "overlap", 
+#'               assay.type = "counts")
+#' reducedDim(tse, "MDS")[1:6, ]
+#'               
+#' ### Unifrac dissimilarity
+#' 
+#' res <- getDissimilarity(tse, method = "unifrac", weighted = FALSE)
+#' dim(as.matrix((res)))
+#' 
+#' tse <- addDissimilarity(tse, method = "unifrac", weighted = TRUE)
+#' reducedDim(tse, "unifrac")[1:6, 1:6]
+#' 
+NULL
+
+#' @rdname getDissimilarity
+#' @export
+setGeneric(
+  "addDissimilarity", signature = c("x"), function(x, method, ...)
+    standardGeneric("addDissimilarity"))
+
+#' @rdname getDissimilarity
+#' @export
+setMethod(
+  "addDissimilarity", signature = c(x = "SummarizedExperiment"),
+  function(
+    x, method, assay_name = "counts", assay.type = assay_name, name = method,
+    transposed = FALSE, tree_name = "phylo", tree.name = tree_name, ...){
```

Comment:
transposed means that the matrix is transposed. This means ultimately that the matrix do not fit anymore to the original TreeSE (dissimilarity matrix should have samples in rows). --> We should remove and disable transposed from addDissimilarity. Catch it so that it is not fed to getDissimilarity()

Reply:
Btw what is the motivation for having "transposed" argument in the first place?

Reply:
It is to allow calculation row-wise (dissimilarity between features). Also, one option is to store dissimilarity matrix to metadata() for instance (however, reducedDim is optimal place for sample-wise dissimilarities).

Reply:
Yes right. 

It is true that reducedDim is quite helpful solution for that. I am somewhat concerned, however, that it might be confusing as a) it can only handle one type of output from dissimilarity function, and b) it is not an actual reduced dimensional presentation (nor altExp). 

How about using a slot with recommended standard name in metadata()?

Reply:
If I understood correctly, I should remove the `transposed` argument from the `getDissimilarity` and `addDissimilarity` functions but should I also make modifications about the storage of dissimilarity matrix or will it be addressed in another PR?

Reply:
I would I would keep `transposed`. Instead of storing the result to reducedDim(), we could store it to `metadata()`.

However, one thing to keep in mind is that metadata() is just a list. So this does not work `metadata(tse, "diss")`, but this works `metadata(tse)[["diss"]]`